runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowMetrics.java

```diff
@@ -206,6 +122,167 @@ public MetricQueryResults queryMetrics(MetricsFilter filter) {
     return result;
   }
 
+  private static class DataflowMetricResultExtractor {
+    ImmutableList.Builder<MetricResult<Long>> counterResults;
+    ImmutableList.Builder<MetricResult<DistributionResult>> distributionResults;
+    ImmutableList.Builder<MetricResult<GaugeResult>> gaugeResults;
+
+    DataflowMetricResultExtractor() {
+      counterResults = ImmutableList.builder();
+      distributionResults = ImmutableList.builder();
+      gaugeResults = ImmutableList.builder();
+    }
+
+    public void addMetricResult(
+        MetricKey metricKey,
+        com.google.api.services.dataflow.model.MetricUpdate committed,
+        com.google.api.services.dataflow.model.MetricUpdate attempted) {
+      if (committed.getDistribution() != null && attempted.getDistribution() != null) {
+        // distribution metric
+        distributionResults.add(
+            DataflowMetricResult.create(metricKey.metricName(), metricKey.stepName(),
+            getDistributionValue(committed), getDistributionValue(attempted)));
+      } else if (committed.getScalar() != null && attempted.getScalar() != null) {
+        // counter metric
+        Long committedLong = getCounterValue(committed);
+        Long attemptedLong = getCounterValue(attempted);
+        counterResults.add(
+            DataflowMetricResult.create(
+                metricKey.metricName(), metricKey.stepName(), committedLong, attemptedLong));
+      } else {
+        // warning
+        LOG.warn("Unexpected metric type. Please report JOB ID to Dataflow Support.");
```

Comment:
How about reporting the metric key? We could also dump the whole update proto, but that may be too much.